### ISSUE TITLE
change the topic

### ISSUE THREAD

**Issue by walide67** [Program-O/Program-O#250]
hello friends 
help me please 
how i can change the topic in aiml ??? 
for e. g 
the bot in the topic number 1 if the bot don't find the answer in this topic change it to another topic ...

**Comment by walide67** [comment]
please help me !!!!

**Comment by Dave-Morton** [comment]
I haven't been ignoring you, @walide67 - I've been ill, and unable to reply.

Setting a topic with AIML is simple. You just use the `<set>` tag, like this:

```xml
<category>
    <pattern>LET S TALK ABOUT *</pattern>
    <template>
        Ok, the subject will be <set name="topic"><star/></set>.
    </template>
</category>
```
In order to _use_ topics in AIML, you need to surround your AIML categories that involve a given topic with `<topic>` tags, like this:

```xml
<topic name="{insert your given topic here}">
    <category>
        <pattern>WHAT IS YOUR FAVORITE</pattern>
        <template>I like Hawaiian pizzas, personally.</template>
    </category>
{more AIML categories about pizza go here}
</topic>
```

Care must be taken when using `<topic>` tags, however. Once a topic is set, only categories within that topic are matched against, which could end up in an endless loop of responses, regardless of the input the user gives. In order to prevent this, a "breakout" category needs to be in place to "change the subject":

```xml
<category>
    <pattern>LET S TALK ABOUT SOMETHING ELSE</pattern>
    <template>
        <think><set name="topic"></set></think>
            Ok, what do you want to talk about?
    </template>
</category>
```

**Comment by walide67** [comment]
i want to breakout from the topic i set like this 
`
<topic name="nametopic">
<category><pattern>somthing from topic</pattern><template>template from topic</template>
</category>
<category><pattern>*</pattern><template>please reforme your question<think><set name="topic"></set></think></template></category>{for breakout from topic }
</topic>`
but if the wildcard is a question from another topic or any question that set in the database the bot must give the good  answer to user

**Comment by Dave-Morton** [comment]
I'm not sure I understand what you mean here, and without seeing your exact AIML code in whole it's difficult to advise you. If you want to email your AIML code to me I can take a look at it and make some suggestions.

**Comment by walide67** [comment]
[password.txt](https://github.com/Program-O/Program-O/files/882586/password.txt)
attache file is my aiml script 
_____________________________________________________
it is a topic about user password to allowed the bot give a Solution to user ...
_______________________________________________________
ok ... suppose that the bot in the password topic  
Then the user send a question that don't set in the topic ... the bot use the wildcard category bellow : 
`<category><pattern>*</pattern><template>
<think><set name="topic"></set></think>
you change the topic  </template></category>` 
ok ... if the wildcard is for example goodbye < i don't set answer for goodbye In all topic > but i have another topic for Salutation 
how i can make the bot  If he does not find the answer in the currently topic Go to search in an other topic  
I hope that the idea arrived :)

**Comment by Dave-Morton** [comment]
the reason why I suggested that you email me the AIML file is to prevent public access to potentially sensitive information. while I can edit your posts here, I cannot remove any attached files without simply deleting your post (which I may still do if you cannot remove the attached file). I've looked at the file, but am so tied up with other things right now that it will likely be the weekend before I can actually do something with it, but I'll get back to you as soon as I can.

**Comment by walide67** [comment]
my aiml script
````
<?xml version="1.0" encoding="UTF-8"?>
<aiml>
<category><pattern>mot de passe</pattern><template>
<think>
<set name="topic">login</set>
</think>
qelle est le problem ??!!
</template></category>
<category><pattern>mot de passe *</pattern><template>
<think>
<set name="topic">login</set>
</think>
qelle est le problem ??!!
</template></category>
<category><pattern>_ mot de passe *</pattern><template>
<think>
<set name="topic">login</set>
</think>
Quel est votre problème avec le mot de passe ??!!... expliquer plus SVP
</template></category>
<category><pattern>* mot de passe</pattern><template>
<think>
<set name="topic">login</set>
</think>
Quel est votre problème avec le mot de passe ??!!... expliquer plus SVP
</template></category>
<topic name="login">
<category><pattern>mot de passe oublié</pattern><template><a href="login/pass_oblier.html">récupération de mot de passe</a></template></category>
<category><pattern>*</pattern><template>
<think><set name="topic"></set></think>
</template></category>
<category><pattern>J'ai oublie mot de passe</pattern><template>tu peut utiliser le lien suivant pour la récupération de votre mot de passe <srai>mot de passe oublié</srai></template></category>
<category><pattern>_ oublié * mot de passe</pattern><template>Vous pouvez utiliser le lien ci-dessous pour rétablir votre mot de passe</template></category>
<category><pattern>oublié *</pattern><template>Si vous avez oublié votre mot de passe Vous pouvez utiliser le lien ci-dessous <srai>mot de passe oublié</srai></template></category>
<category><pattern>comment * mot de passe</pattern><template>Vous pouvez utiliser le lien ci-dessous pour rétablir votre mot de passe <srai>mot de passe oublié</srai></template></category>
<category><pattern>mot de passe</pattern><template>vous avez oublie votre mot de passe</template></category>
<category><pattern>oui</pattern><that>vous avez oublie votre mot de passe</that><template>tu peut utiliser le lien suivant pour la récupération de votre mot de passe <srai>mot de passe oublié</srai></template></category>
<category><pattern>no</pattern><that>vous avez oublie votre mot de passe</that><template>Quel est le probleme alors ?</template></category>
<category><pattern>comment changer mot de passe</pattern><template><script>chemin_bot()</script></template></category>
<category><pattern>comment _ changer mot de passe</pattern><template><srai>comment changer mot de passe</srai></template></category>
<category><pattern>changer mot de passe</pattern><template><srai>comment changer mot de passe</srai></template></category>
<category><pattern>email</pattern><template></template></category>
</topic>
</aiml>
````

**Comment by walide67** [comment]
Thank you for your efforts @Dave-Morton

**Comment by AlekzNet** [comment]
@walide67 I notice this in your AIML:

```xml
<pattern>J'ai oublie mot de passe</pattern>
```

I would remove the `'` and rewrite it like this:

```xml
<pattern>J ai oublie mot de passe</pattern>
```

**Comment by AlekzNet** [comment]
@walide67 ... and reformatted your AIML, so it's easier to understand (one of the reasons I love python ;) )

**Comment by walide67** [comment]
and what i ca do if user use ' in his question (if user write : j'ai oublie mot de passe ?? ) @AlekzNet

**Comment by Dave-Morton** [comment]
All punctuation is stripped from the user's input, and the text is "normalized", so Program O will never try to match it. So if the user enters `j'ai oublie mot de passe`, Program O will try to match against `J AI OUBLIE MOT DE PASSE`. This is done automatically, so there is nothing that needs to be done. when writing your AIML, you should _never_ use punctuation in any pattern-side tags (e.g. `PATTERN`, `THAT` or `TOPIC`), and you _should_ use uppercase text to improve performance.

**Comment by walide67** [comment]
what about topic 
how i can get a answer from outside the topic

**Comment by AlekzNet** [comment]
Until you change on unset  the topic, you will not get an answer outside it.

**Comment by Dave-Morton** [comment]
All topics need some way to leave it. This is called a "breakout" category, and the most simple and effective one looks like this:

```xml
<category>
    <pattern>*</pattern>
    <template><think><set name="topic></set></think><sr/></template>
</category>
```
This category unsets the topic, then uses the `<sr>` tag, which is shorthand for `<srai><star/></srai>`.

**Comment by walide67** [comment]
what about the answer of ''hello'' for example or the General questions,  I put her answers in all topic ??

**Comment by walide67** [comment]
what happens if the wildcard (</sr>)  Does not exist in AIML files @Dave-Morton

**Comment by Dave-Morton** [comment]
I really need to make time soon to write a wiki page here on AIML. {sigh!}

@walide67 while I'm certainly willing to answer support questions here, and even the _occasional_ non-support question. the questions you've been asking recently are really outside the scope of the issues section of the project. I strongly suggest that you do a couple of things here:

1. Try very hard to do some outside research on how AIML works. there are several excellent sources available on the internet on how to create AIML files. One such source is [chatbots.org](https://chatbots.org). there is a rather large community of people there who are passionate about all areas of chatbots and AI, and would be more than happy to guide you and answer your questions. I'm an administrator there, and would be happy to answer AIML questions on those forums.
2. Try to limit questions here to technical aspects of Program O itself. When you ask questions here that aren't directly related to Program O, it makes it harder for others to find answers to their questions, and it takes the developers away from other things that they might be doing.

The bottom line here is that questions like these don't really belong in the issues section (partly because they aren't really issues), and should be asked elsewhere. I'm more than willing to answer them. Just not here.

**Comment by walide67** [comment]
I understand you, Thank you so much @Dave-Morton  @AlekzNet ... and i am sorry

**Comment by Dave-Morton** [comment]
No need to apologize. I'm sure you weren't aware.
